### PR TITLE
Multiple modules

### PR DESCRIPTION
**Pull request by bbqbaron**
From #15. This handles modules-within-modules according to the following logic. Certainly open to feedback, since I haven't written enough flow decls to know all the permutations!

* Nested modules names are parsed as delimited by `/`
* Any level of nesting after the first is considered a single submodule name: e.g. `foo/bar/baz` becomes a module `BarBaz` in the `foo.re` output. I could also recursively nest them by ensuring a `Bar` and adding the `Baz` decls inside it; not sure if this is a use case that happens a lot. I wanted to get an impression/opinion before adding what seems like a polish feature.
* Externals are each declared with their full module path, inside their declared modules. That is, I `@@bs.module "foo/bar/whatever"` inside each external in `Foo.Bar`.

Some changes are formatting; will annotate the one real change.

### CODE REVIEW

File: src/retyped/codegen.re [hunk 1]
```diff
@@ -22,19 +22,14 @@ let rec bstype_name =
   | Function _ => "func"
```
Comment (bbqbaron) [inline src/retyped/codegen.re]: Formatting from a `refmt` run ensues; skip until next comment.

File: src/retyped/codegen.re [hunk 2]
```diff
@@ -346,22 +307,35 @@ let rec declaration_to_code module_id types =>
...
-      name::(String.uncapitalize_ascii id)
-      type_of::(bstype_to_code type_of)
-      ();
+      name::(String.uncapitalize_ascii id) type_of::(bstype_to_code type_of) ();
 
 let program_to_code program =>
   switch program {
   | ModuleDecl id statements =>
     let typeof_table = Typetable.create statements;
+    /* is the module nested ? */
```
Comment (bbqbaron) [inline src/retyped/codegen.re]: Here we go:

* look for `/`-delimited name
* create `module Whatever = {`/`}` wrapper and apply to code if present

File: src/retyped/jbuild [hunk 3]
```diff
@@ -2,4 +2,4 @@
 
 (library
   ((name retyped)
-    (libraries (sedlex flow_parser))))
+    (libraries (js_of_ocaml sedlex flow_parser))))
```
Comment (bbqbaron) [inline src/retyped/jbuild]: This is to get access to `Regexp` in jsoo. It's already used in `src`, so this should be fine, right?

Comment (rrdelaney) [inline src/retyped/jbuild]: Unfortunately we can't compile to native if `src/retyped` depends on jsoo 😕

screen shot 2017-07-16 at 2 13 19 pm

Is there some other Regex library or implementation we could use?

**Comment by rrdelaney** [comment]
Ok, so first off this is seriously amazing work 🎉 This is honestly fantastic!

The PR looks great overall! The only problem is having `src/retyped` take a dependency of jsoo. Right now it's compiled as a library dependency of the two modules in`src`, meaning it can target native OCaml if you run `jbuilder build src/cli.exe` - hopefully we don't have to lose that.

Can we take a dependency on another OCaml regex library? Native target is really helpful for development speed.

You can test out the native target with `jbuilder build src/cli.exe` and then run `./_build/default/src/cli.exe -debug <my-file.js>`

File: src/retyped/jbuild (hunk 3 above)
Comment (bbqbaron) [inline src/retyped/jbuild]: Okay! So after some experimentation I just dropped regexes and wrote my own `split` function, which at least held up to the testing from the use case. `Str` had unimplemented primitives in JS, and of course `Regexp` has them in native, so eventually I just called it. I even considered [String](https://caml.inria.fr/pub/docs/manual-ocaml/libref/String.html)'s `split_on_char`, but it's 4.04.0+, and isn't Reason not there yet?

```
A developer with problem once said, "I know! I'll use regular expressions!" Now he has two problems.

--Unknown
```

Comment (rrdelaney) [inline src/retyped/jbuild]: I don't think we can use stuff from 4.04 yet 😐 But implementing `split` was perfect 😄

**Comment by rrdelaney** [comment]
This is perfect! Thanks for implementing this so quickly 😄

**Comment by bbqbaron** [comment]
I'd say so; native and JS tests seem to work (where "native test" is just building `cli.exe` and running it on my test file).

**Pull request merged**